template.go

```diff
@@ -0,0 +1,437 @@
+	return templateDocument(f, sections, []string{"template", "email"})
+}
+
+// function to build the function map for the template engine from the global faker
+func createGlobalFakerFunctionMap() template.FuncMap {
+	funcMap := template.FuncMap{}
+
+	// functions that wont work with template engine
+	incompatible := map[string]string{
+		"ShuffleAnySlice": "ShuffleAnySlice",
+		"ShuffleInts":     "ShuffleInts",
+		"ShuffleStrings":  "ShuffleStrings",
+		"Slice":           "Slice",
+		"Struct":          "Struct",
+	}
+
+	// Build the function map from the globalFaker
+	v := reflect.ValueOf(globalFaker)
+	// loop through the methods
+	for i := 0; i < v.NumMethod(); i++ {
+		method := v.Method(i)
+		// get the method name lowercase
+		method_name := v.Type().Method(i).Name
+
+		//check if method_name is in the incompatable list
+		if _, ok := incompatible[method_name]; ok {
+			continue
+		}
+
+		funcMap[method_name] = method.Interface()
+	}
+
+	// add the custom functions
+
+	// function to generate a range of integers
+	funcMap["IntRange"] = func(start, end int) []int { // function to generate a range of integers
+		n := end - start + 1
+		result := make([]int, n)
+		for i := 0; i < n; i++ {
+			result[i] = start + i
+		}
+		return result
+	}
+
+	// function to generate a range of integers and store in intRangeResult
+	funcMap["CreateListResult"] = func(start, end int) intRangeResult {
+		n := end - start + 1
+		result := make([]int, n)
+		for i := 0; i < n; i++ {
+			result[i] = start + i
+		}
+		return intRangeResult{Value: start - 1, Range: result}
+	}
+
+	// function to remove value from list
+	funcMap["ListResult"] = func(r intRangeResult, min int, max int, shuffle bool) intRangeResult {
+		var new_list []int
+
+		// if the remove value is in the list remove it
+		if len(r.Range) > 0 {
+			for _, v := range r.Range {
+				if v != r.Value {
+					new_list = append(new_list, v)
+				}
+			}
+		}
+
+		// See if the list is empty
+		if len(r.Range) == 0 {
+			// create a new slice of ints
+			n := max - min + 1
+			for i := 0; i < n; i++ {
+				new_list = append(new_list, min+i)
+			}
+		}
+
+		// Randomize the slice
+		if shuffle {
+			shuffleInts(globalFaker.Rand, new_list)
+		}
+		// return the first value and the rest of the list
+		return intRangeResult{Value: new_list[0], Range: new_list[1:]}
+	}
+
+	// function to generate a base64 encode string useful for images
+	funcMap["Base64Enc"] = base64EncString
+
+	// function  to replace all values in string
+	funcMap["Replace"] = strings.ReplaceAll
+
+	// function to make string lower case
+	funcMap["lc"] = strings.ToLower
+
+	// function to concatenate strings
+	funcMap["conc"] = func(args ...string) string {
+		return strings.Join(args, " ")
+	}
+	// function to make string upper case
+	funcMap["uc"] = strings.ToUpper
+
+	// function wrapper for SVG this is because template engine cant handle passing structs
+	funcMap["SVG"] = func(width int, height int) string {
```

Comment:
remove for now. Cant just use globalFaker cause we need to use the rand that was used in the function request.